### PR TITLE
reengineering: implement core components for online algorithms

### PR DESCRIPTION
**Pull request by Desiment**
# PySATL-CPD reengineering: step I - online algorithms

## Context
The PySATL-CPD library contains a lot of code with significant architectural debt, unclear abstractions, and inconsistent patterns across the codebase. The existing implementation suffered from unnecessary abstractions, incomplete benchmarking, limited visualization capabilities and incorrectness from mathematical standpoint. 

I suggest to start from scratch. This PR is first in series of PR during which new architecture will be developed, and existing algorithms will be re-implemented with focus on correctness. As one can see, first commit in this branch effectively removes almost all codebase. 

## Overview

This PR introduces the foundational infrastructure for online change-point detection. It establishes a framework for implementing online detection algorithms, executing detection loops, and managing detection results. The implementation includes a complete Shewhart control chart algorithm, comprehensive test infrastructure with mocks, and full mypy strict mode compliance.

### 1. Typing
- **`typedefs.py`**: Core type aliases (`Number`, `UnivariateNumericArray`, `MultivariateNumericArray`, `BoolArray`, etc) providing type safety across the library

- Generic type parameters for maximum flexibility and reusability

### 2. Data Providers Module
- **`DataProvider[T]`**: Abstract interface for sequential data access
- **`NDArrayUnivariateProvider`**: Wrapper for 1D NumPy arrays (univariate time series)
- **`NDArrayMultivariateProvider`**: Wrapper for 2D NumPy arrays (multivariate time series)

### 3. Detection Trace Framework
- **`DetectionTrace[T]`**: Base container for detection results with change point indices and observation scores
- **`OnlineDetectionStepResult`**: Per-step results including detection flags, statistics, and timing
- **`OnlineDetectionTrace[StateT]`**: Complete trace for online detection with:
  - Detection function values
  - Processing times
  - Algorithm state snapshots
  - Skip periods and forced change points tracking
  - Factory method for constructing from step results

### 4. Online Algorithm Architecture
- **`OnlineAlgorithm[T, C, S]`**: Generic abstract base class with:
  - Configuration management (immutable)
  - State snapshots (immutable)
  - Process and reset methods
  - Name property for identification
- **`OnlineAlgorithmState`**: Base immutable state class
- **`OnlineAlgorithmConfiguration`**: Base immutable configuration class

### 5. Shewhart Control Chart Implementation
- Complete online algorithm with:
  - Sliding window statistics
  - Running mean and variance using Welford's algorithm
  - Learning period support
  - Proper state and configuration dataclasses
  - Comprehensive validation for input parameters

### 6. Online Detection Solver
- **`OnlineCpdSolver[T, C, S]`**: execution engine with:
  - Sequential observation processing
  - Threshold-based detection
  - Configurable skip periods after detections
  - Maximum run-length forced detections
  - Algorithm reset on detection
  - Timing measurement per step
  - Input validation for all parameters
  - The main method `run` now accepts an algorithm, a DataProvider, and a threshold. This way, a single instance of OnlineCpdSolver can be used on multiple providers or algorithms

### 7. Comprehensive Test Suite
- **Mock Infrastructure**:
  - `MockUnivariateDataProvider`, `MockMultivariateDataProvider` and various variations containing: inf, nan, zero and constant
  - `MockEmptyDataProvider`, `MockSingleObservationProvider`
  - `MockUnivariateDirtyDataProvider`, `MockMultivariateDirtyDataProvider`
  - `MockOnlineAlgorithm` with configurable behavior
  - State sequence and return sequence support
- **Unit Tests**:
  - Data providers (abstract and concrete implementations)
  - Detection trace classes
  - Online algorithm interface
  - Shewhart control chart algorithm
  - Online CPD solver with edge cases
  - Labeled data container

## Future Work
- Offline algorithm infrastructure
- Additional online algorithms (CUSUM, EWMA)
- Visualization module implementation
- Benchmarking suite
- Documentation website

### CODE REVIEW

File: pysatl_cpd/core/online/online_cpd_solver.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+        # Validate max_runlength is positive if specified
+        if max_runlength is not None and max_runlength <= 0:
+            raise ValueError(f"max_runlength must be positive if specified, got {max_runlength}")
+
+        self.__skip_period = skip_period
+        self.__max_runlength = max_runlength
+        self.__collect_states = collect_states
+
+    def run[DataT, ConfugrationT: OnlineAlgorithmConfiguration, StateT: OnlineAlgorithmState](
```
Comment (iraedeus) [inline pysatl_cpd/core/online/online_cpd_solver.py]:
```suggestion
    def run[DataT, ConfigurationT: OnlineAlgorithmConfiguration, StateT: OnlineAlgorithmState](
```

File: pysatl_cpd/algorithms/online/shewhart_control_chart.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+            algorithm._mean = state.mean
+            algorithm._variance = state.variance
+            algorithm._standard_deviation = state.standard_deviation
+            algorithm._samples_count = state.samples_count
+            algorithm._window = deque(state.window_contents, maxlen=configuration.window_size)
+            algorithm._window_sum = state.window_sum
+            algorithm._window_mean = state.window_mean
+            # Recalculate previous_mean for variance update
+            if algorithm._samples_count > 0:
+                algorithm._previous_mean = algorithm._mean - (algorithm._mean - algorithm._previous_mean)
```
Comment (iraedeus) [inline pysatl_cpd/algorithms/online/shewhart_control_chart.py]: Unnecessary code: the `process` method already overwrites `self.__previous_mean` before calculating variance, using the `_update_mean` method

**Review by iraedeus — CHANGES_REQUESTED** [review]

File: pysatl_cpd/core/online/online_detection_trace.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+            np.array([step.processing_time for step in steps], dtype=np.float64),
+        )
+
+        # Extract algorithm states preserving None values
+        states: list[StateT | None] = [step.algorithm_state for step in steps]
+        skip_periods = extract_periods([s.is_in_skip_period for s in steps])
+        learning_periods = extract_periods([s.is_in_learning_period if s is not None else None for s in states])
+
+        # Identify indices of different detection types
+        detected_indices: list[int] = [idx for idx in step_nums if steps[idx].is_change_point]
```
Comment (iraedeus) [inline pysatl_cpd/core/online/online_detection_trace.py]:
```suggestion
        detected_indices = [step.step_num for step in steps if step.is_change_point]
```

File: pysatl_cpd/core/online/online_detection_trace.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        forced_indices: list[int] = [idx for idx in step_nums if steps[idx].is_forced_change_point]
```
Comment (iraedeus) [inline pysatl_cpd/core/online/online_detection_trace.py]:
```suggestion
        forced_indices = [step.step_num for step in steps if step.is_forced_change_point]
```

File: pysatl_cpd/core/online/online_detection_trace.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        signal_indices: list[int] = [idx for idx in step_nums if steps[idx].is_signal_change_point]
```
Comment (iraedeus) [inline pysatl_cpd/core/online/online_detection_trace.py]:
```suggestion
        signal_indices = [step.step_num for step in steps if step.is_signal_change_point]
```

**Review by iraedeus — CHANGES_REQUESTED** [review]

**Review by iraedeus — APPROVED** [review]
LTGM
